### PR TITLE
platform/evdev: Use ConsoleServices

### PR DESCRIPTION
**Pull request by RAOF**
This migrates the libinput platform to the use of `ConsoleServices` for device access.

### CODE REVIEW

**Comment by RAOF** [comment]
Marked as no-merge: this works locally, but there's some obvious factoring to be done in the libinput tests. I'm pushing this up now to get a CI run done on it.

**Comment by RAOF** [comment]
Review guide:
This includes the bridge for an annoying impedance mismatch - libinput expects to you to synchronously provide a file descriptor for a device node, while logind will *a*synchronously provide you with a file descriptor for the device node with a given devnum.

To resolve this, we switch from having libinput handle the udev events internally to handling them ourselves. This allows us to request the device from logind when we receive an ADD event, store it in the `FdStore`, and *then* call `libinput_path_add_device` to create the libinput object - at which point libinput will request the fd which we can provide from the `FdStore`.

There are a number of gotchas in the test environment, but these should all have comments in the code explaining the oddness.

File: src/platforms/evdev/fd_store.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright © 2017 Canonical Ltd.
+ *
```
Comment (AlanGriffiths) [inline src/platforms/evdev/fd_store.cpp]: Is the year correct?

File: src/platforms/evdev/libinput_ptr.cpp [hunk 2]
```diff
@@ -26,25 +27,25 @@
...
+    auto fd_store = static_cast<mie::FdStore*>(userdata);
+
+    auto fd = fd_store->take_fd(path);
+
+    if (fcntl(fd, F_SETFL, flags) == -1)
     {
-        int const clockId = CLOCK_MONOTONIC;
-        // Switch each evdev 'client' to MONOTONIC
-        ioctl(evdev, EVIOCSCLOCKID, &clockId);
+        std::cerr << "Failed to set flags: " << strerror(errno) << " (" << errno << ")" << std::endl;
```
Comment (AlanGriffiths) [inline src/platforms/evdev/libinput_ptr.cpp]: Better as a log_warning()?

**Comment by AlanGriffiths** [comment]
For those that are following impatiently, the remaining change to activate the logind support is:
```
diff --git a/src/server/console/default_configuration.cpp b/src/server/console/default_configuration.cpp
index 06ebc1ec3e..e3c092d834 100644
--- a/src/server/console/default_configuration.cpp
+++ b/src/server/console/default_configuration.cpp
@@ -17,6 +17,7 @@
  */
 
 #include "mir/default_server_configuration.h"
+#include "mir/glib_main_loop.h"
 #include "mir/options/configuration.h"
 #include "mir/log.h"
 #include "mir/emergency_cleanup.h"
@@ -97,21 +98,31 @@ std::shared_ptr<mir::ConsoleServices> mir::DefaultServerConfiguration::the_conso
     return console_services(
         [this]() -> std::shared_ptr<ConsoleServices>
         {
-            try
-            {
-                auto const vt_services = std::make_shared<mir::LinuxVirtualTerminal>(
-                    std::make_unique<RealVTFileOperations>(),
-                    std::make_unique<RealPosixProcessOperations>(),
-                    the_options()->get<int>(options::vt_option_name),
-                    *the_emergency_cleanup(),
-                    the_display_report());
-                mir::log_debug("Using Linux VT subsystem for session management");
-                return vt_services;
-            }
-            catch (...)
-            {
-                mir::log_debug("No session management supported");
-                return std::make_shared<mir::NullConsoleServices>();
-            }
+                try
+                {
+                    auto const vt_services = std::make_shared<mir::LogindConsoleServices>(
+                        std::dynamic_pointer_cast<GLibMainLoop>(the_main_loop()));
+                    mir::log_debug("Using logind for session management");
+                    return vt_services;
+                }
+                catch (...)
+                {
+                    try
+                    {
+                        auto const vt_services = std::make_shared<mir::LinuxVirtualTerminal>(
+                            std::make_unique<RealVTFileOperations>(),
+                            std::make_unique<RealPosixProcessOperations>(),
+                            the_options()->get<int>(options::vt_option_name),
+                            *the_emergency_cleanup(),
+                            the_display_report());
+                        mir::log_debug("Using Linux VT subsystem for session management");
+                        return vt_services;
+                    }
+                    catch (...)
+                    {
+                        mir::log_debug("No session management supported");
+                        return std::make_shared<mir::NullConsoleServices>();
+                    }
+                }
         });
 }
```

**Pull request merged**